### PR TITLE
add train test split

### PR DESCRIPTION
**Pull request by bishwarup307**
## What this PR is for?
To introduce splitting datasets into train and test (validation)

### 1. add train test split
```
annotation = Annotation(root = /path/to/annotation/dir, format = "coco") # or yolo, pascal, etc.
splits = annotation.split(test_size = 0.2, stratified = True, random_state = 0)
```
The`splits` above in itself is a `FormatSpec` instance and as such all methods on top of it e.g., `show_distribution`, `bbox_scatter` etc. works out of the box. One can save the split simply as:
```
splits.save(output_dir = "~/Desktop")
```
while saving one can also pass an additional `export_to` format and the split dataset will be exported to that format. Please note the `copy_images` is set to `True` by default in `save`.

### 2. Additional column in `master_df` named `image_path`.
The reason is while trying to copy the images after splitting it is really hard to maintain the mapping of images to its location in disk with just `_has_image_split` (an  user may have already a split dataset but wants to "stratify" the data for example).
With these change, the path of each image is recorded correctly in the `master_df` itself so that it doesn't need to be resolved for any downstream tasks. `copyfile` and `_fastcopy` also modified to reflect this change in copying images.
> Only implemented for "coco" source format in this PR

### 3. Make the exports directory layout more consistent. 
At present the export directory layout is somewhat messed up. Made the directory structure more consistent.
- if `save_under` is not specified, it defaults to `to` annotation format.
- if no `output_dir` is specified, it defaults to `root`
- for `to` formats where there is no individual file required for each image (e.g., `coco`, `sagemaker`) the annotation is saved under `output_dir/{to}/annotations/{split}.json(.manifest/.csv)`. For the formats where individual annotations are desired e.g., `yolo`, `pascal`, the annotations will be in `output_dir/{to}/annotations/{split}/*.txt(.xml)`
- if `copy_images` is `True`, the images are in `output_dir/{to}/images/{split}`

### 4. Remove abstractclass from  FormatSpec
Otherwise `train_test_split` cannot return the same class instance as it has a abstractmethod `_resolve_dataframe`.
## What's next?
- Need to make the output directory layout for tfrecord (can do after merge)
- Need to modify `_resolve_dataframe` for all other formats besides `coco`

### CODE REVIEW

File: optical/__init__.py [hunk 1]
```diff
@@ -1 +1 @@
-__version__ = "0.0.1"
+__version__ = "0.0.11"
```
Comment (bishwarup307) [inline optical/__init__.py:1]: This is unintended lol. ignore this for the moment!

File: .vscode/settings.json [hunk 2]
```diff
@@ -27,4 +27,14 @@
...
+    "workbench.colorCustomizations": {
+        "activityBar.activeBackground": "#ab307e",
+        "activityBar.activeBorder": "#25320e",
+        "activityBar.background": "#ab307e",
+        "activityBar.foreground": "#e7e7e7",
+        "activityBar.inactiveForeground": "#e7e7e799",
+        "activityBarBadge.background": "#25320e",
+        "activityBarBadge.foreground": "#e7e7e7"
+    },
+    "python.pythonPath": "/Users/bishwarup.b/opt/anaconda3/envs/optical2/bin/python",
```
Comment (satheeshkatipomu) [inline .vscode/settings.json:39]: Please remove the local python path setting in vscode settings.

Comment (bishwarup307) [reply .vscode/settings.json:39]: zzz, this happens everytime. I have no idea why. Maybe cause i am switching between container and local. Will fix.

**Comment by bishwarup307** [comment]
Please note the 56 commits are due to the fact that the git history was overwritten to remove the `.vscode/settings.json` file from the repo. it changes the commit hash for the previous commits and that's the reason the number of commits doubled. It's assumed harmless and expected in case we want to [recover the previous state at a future point](https://stackoverflow.com/questions/23820240/git-filter-branch-gave-me-two-sets-of-commits)

**Review by satheeshkatipomu — APPROVED** [review]
Looks good to me.

File: optical/converter/__init__.py [hunk 3]
```diff
@@ -47,23 +47,32 @@ def bbox_scatter(self, split: Optional[str] = None, category: Optional[str] = No
     def bbox_stats(self, split: Optional[str] = None, category: Optional[str] = None):
         return self.formatspec.bbox_stats(split, category)
 
-    def convert(self, to: str, output_dir: Optional[Union[str, os.PathLike]] = None, **kwargs):
+    def export(self, to: str, output_dir: Optional[Union[str, os.PathLike]] = None, **kwargs):
```
Comment (aman0044) [inline optical/converter/__init__.py:50]: Looks better.

**Review by aman0044 — APPROVED** [review]
Looks fine...

**Review by bharatkumarreddy — APPROVED** [review]
Looks good

**Pull request merged**